### PR TITLE
KAFKA-12323: Set timestamp in record context when punctuate

### PR DESCRIPTION
**Pull request by guozhangwang**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@ableegoldman @vvcephei for reviews, thanks!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -792,7 +790,16 @@ public void punctuate(final ProcessorNode<?, ?, ?, ?> node,
...
 
-        updateProcessorContext(node, time.milliseconds(), null);
+        // when punctuating, we need to preserve the timestamp (this can be either system time or event time)
+        // while other record context are set as dummy: null topic, -1 partition, -1 offset and empty header
+        final ProcessorRecordContext recordContext = new ProcessorRecordContext(
+            timestamp,
+            -1L,
+            -1,
+            null,
+            new RecordHeaders()
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:800]: Why not pass `null` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 2]
```diff
@@ -1874,6 +1873,66 @@ public void close() {}
         assertEquals(2, punctuatedWallClockTime.size());
     }
 
+    @Test
+    public void shouldPunctuateWithTimestampPreservedInProcessorContext() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:1876]: I think we should add one more test for `STREAM_TIME` punctuation?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 3]
```diff
@@ -1874,6 +1873,66 @@ public void close() {}
...
+        thread.rebalanceListener().onPartitionsAssigned(assignedPartitions);
+
+        thread.runOnce();
+        assertEquals(0, peekedContextTime.size());
+
+        mockTime.sleep(100L);
+        thread.runOnce();
+
+        assertEquals(1, peekedContextTime.size());
+        assertNotNull(peekedContextTime.get(0));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: We should verify the actual timestamp.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR.

Can we also update the JavaDocs of `ProcessorContext#forward(key, value)`? It currently says:
```
Used the input record's timestamp as timestamp for the output record.
```

We should explain how it works when used in a punctuation.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:800]: I'm following the existing behavior here: if the record context is `null`, we also return a `return new RecordHeaders();`.

**Review by guozhangwang — COMMENTED** [review]
Addressed comments.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 4]
```diff
@@ -1805,18 +1806,16 @@ public void shouldPunctuateActiveTask() {
         final List<Long> punctuatedStreamTime = new ArrayList<>();
         final List<Long> punctuatedWallClockTime = new ArrayList<>();
         final org.apache.kafka.streams.processor.ProcessorSupplier<Object, Object> punctuateProcessor =
-            () -> new org.apache.kafka.streams.processor.Processor<Object, Object>() {
+            () -> new org.apache.kafka.streams.processor.AbstractProcessor<Object, Object>() {
                 @Override
                 public void init(final org.apache.kafka.streams.processor.ProcessorContext context) {
                     context.schedule(Duration.ofMillis(100L), PunctuationType.STREAM_TIME, punctuatedStreamTime::add);
                     context.schedule(Duration.ofMillis(100L), PunctuationType.WALL_CLOCK_TIME, punctuatedWallClockTime::add);
+                    context.schedule(Duration.ofMillis(200L), PunctuationType.WALL_CLOCK_TIME, timestamp -> context.forward("key", "value"));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: What is the purpose of adding this line? Seem this case is covered in the newly added test below?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 5]
```diff
@@ -1874,6 +1873,85 @@ public void close() {}
...
+             ConsumerRecord.NULL_CHECKSUM,
+             "K".getBytes().length,
+             "V".getBytes().length,
+             "K".getBytes(),
+             "V".getBytes()));
+
+        thread.runOnce();
+
+        assertEquals(2, peekedContextTime.size());
+        assertEquals(0L, peekedContextTime.get(1).longValue());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:1951]: Why is the expected value `0L` ? Should it not be `100L` (ie, the input record's timestamp that dictates stream-time)?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: No it is not intentional, removed.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:1951]: It is not actually.. when we schedule at init we always add the scheduled event with the current timestamp

```
case STREAM_TIME:
                // align punctuation to 0L, punctuate as soon as we have data
                return schedule(0L, interval, type, punctuator);
            case WALL_CLOCK_TIME:
                // align punctuation to now, punctuate after interval has elapsed
                return schedule(time.milliseconds() + interval, interval, type, punctuator);
```

hence for stream time punctuation, the first one always triggers right when we received the first record.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.8 / 2.7, cc RM @vvcephei

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 5 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:1951]: So to make sure we actually use "stream-time" we should change the test to actually punctuation twice? Would you mind doing a follow-up PR? Might also be worth not use a different timestamp compare to "wall-clock time" to make sure we don't by accident pass in current wall-clock.

**Pull request merged**